aiohttp/client_ws.py

```diff
@@ -235,11 +239,32 @@ async def receive(self, timeout: Optional[float] = None) -> WSMessage:
 
             try:
                 self._waiting = self._loop.create_future()
+
+                read_task: asyncio.Task[WSMessage] = asyncio.create_task(
+                    self._reader.read()
+                )
+                is_heartbeat_failed_task: asyncio.Task[bool] = asyncio.create_task(
+                    self._is_heartbeat_failed.wait()
+                )
                 try:
                     async with async_timeout.timeout(
                         timeout or self._timeout.ws_receive
                     ):
-                        msg = await self._reader.read()
+                        # Check the heartbeat status when waiting data from server
+                        done, pending = await asyncio.wait(
+                            (read_task, is_heartbeat_failed_task),
+                            return_when=asyncio.FIRST_COMPLETED,
+                        )
+                        # If server doesn't pong, but return data normally,
+                        # supress the exception.
+                        if read_task in done:
+                            is_heartbeat_failed_task.cancel()
+                            msg = read_task.result()
+                        elif is_heartbeat_failed_task in done:
+                            read_task.cancel()
+                            assert self._exception is not None
+                            raise self._exception
```

Comment:
I'm not sure that raise ` TimeoutError` or just return a `WSMessage(WSMsgType.ERROR, ..)` message which is better.